### PR TITLE
Document negative saturation

### PR DESCRIPTION
**Pull request by nauta-turbidus**
Improved the description of the `saturation` field in the table argument to `player:set_lighting()`. Described how the value is in fact applied. Described what different values do.

Negative values are described to alter the hue severely. What this actually means is actual hue inversion in about 1/3 of cases (180 deg change), 60deg change in another 1/3 of cases, and something in between in the rest of cases (somewhat uniformly distributed).

We're using this behaviour in VoxeLibre, and I've noticed it's undocumented.

## To do

This PR is Ready for Review.

## How to test

Try setting saturation to a negative value in game.

Or use this C++ code to test out the results of the operation on a random sample:
https://gist.github.com/nauta-turbidus/60950e069b15094e08c6678a24af45e1

<details><summary>Screenshot of feature in use:</summary>
<p>
Normal saturation:

screenshot_20240907_011931

Same shot with `saturation == -1`:

screenshot_20240907_011944

</p>
</details>

### CODE REVIEW

**Comment by Zughy** [comment]
I had no idea this thing was possible, and I also think it shouldn't be. Parameter is called "saturation", I don't expect it as a modder to create an invert colours effect

**Comment by nauta-turbidus** [comment]
I could agree to making a second setting that inverts colors and clamping this one, but that would mean:
* breaking API compatibility
* adding additional code under the hood that would do the same thing that the current one already does, in the same way as it already does it
* all the above only to make the API "make more sense"

**Comment by Zughy** [comment]
> breaking API compatibility

There is no breaking compatibility as it's not documented, which is why I'm against this change. VoxeLibre is relying on undocumented behaviour

**Comment by nauta-turbidus** [comment]
Fine, still why add more code that will still be doing the same thing under the hood?

Adding a bool like `invert_colors` to set_lighting() that will be just multiplying `saturation` by -1 under the hood? Fine by me, I can PR it, but why?

Once the consensus is that we're splitting this, I'm PRing this btw.

**Comment by SmallJoker** [comment]
I don't know why you would want to invert the colors in the first place, but as long negative values do not cause bugs I think it's fine to document and support this case. It's great to see when a feature turns out to support more use-cases than originally anticipated.

**Comment by nauta-turbidus** [comment]
The usage is obviously a trippy "something is wrong" effect.

To see what we're doing with it in VL, use `/effect nausea 10` for example.

**Comment by lhofhansl** [comment]
I think this is unintentional.
If that's a feature we want we can document it, otherwise we should probably just block anything <0 or clamp it to 0.

**Comment by teknomunk** [comment]
It may be unintentional, but it's already in use. So, I see four main options:

1. Document the existing behavior (this PR).
2. Create a new parameter that inverts colors, and document that instead.
3. Do nothing. The existing behavior remains undocumented.
4. Remove the behavior entirely, breaking existing game behavior and provide no replacement.

**Comment by Desour** [comment]
Nan, inf, and -inf (f32) cause a black scene.

---

Bad documentation means that there's an intended feature that's not properly documented.
Calling some unspecified behaviour a badly documented feature is like shitting in your neighbors toilet when they left the front door open and saying they just forgot to invite you to come over for shitting.

Anyway, I don't care at all if we add this as feature or remove the unspecified behaviour. I just like abstruse comparisons.

**Comment by appgurueu** [comment]
Relevant code:

https://github.com/minetest/minetest/blob/master/client/shaders/second_stage/opengl_fragment.glsl#L84-L90

Let $c$ be a component of the color, $b$ be the brightness, $s$ be the saturation. This computes $(1 - s)b + sc$.

Now suppose we have a negative saturation $s' = -s$. We now compute $(1 - s')b + s'c = (1 + s)b - sc$.

I'm not sure what you mean by "invert the colors". I see two ways to understand it: First the saturation is applied, then the colors are inverted, or the other way around. But is that actually the case?

Inverting after applying saturation would result in $1 - ((1 - s)b + sc) = (s - 1)b - sc + 1 \neq (1 + s)b - sc$ in general.

Inverting before applying saturation would yield $(1 - s)b + (1 - c)s \neq (1 + s)b - sc$ in general.

In other words, it surely computes *something*, but that something doesn't seem to be just an inversion of colors. Hence I'd be in favor of disallowing this. A more well-behaved color inversion feature could be added in place of it.

**Comment by Zughy** [comment]
So, nausea was added in VL 3 months ago, without reporting the undocumented behaviour
https://git.minetest.land/VoxeLibre/VoxeLibre/commit/aa4ff1827d01a2466998ceba9da3249135e6b88a

after that, 3 VL releases have seen the light
https://git.minetest.land/VoxeLibre/VoxeLibre/tags

And now, when we tell you that it might not be worth it, some of you try to pin it on Minetest, using the "VL is popular" card . Let me be very clear: that's not how it works and I personally don't tolerate this attempt of putting pressure on us with underhand mind games. That's how it should work instead:
1. people notice something undocumented
2. to be sure, people open an issue to upstream (MT) to at least discuss this thing
3. thing is discussed in upstream (VL  & MT)
4. possibly a solution is found and everyone is happy: also future modders, who'll benefit from the cooperation between VL and MT

That's what I do with my mods and my server, so I don't see why you can't do it as well

**Comment by rudzik8** [comment]
> some of you

It would be much better if you removed that school teacher attitude and addressed me directly, the same way I addressed you. You're clearly trying to be respectful here, but that "I won't be pointing my finger at it, but ..." doesn't come across as such.

> Let me be very clear: that's not how it works and I personally don't tolerate this attempt of putting pressure on us when you didn't look for a dialogue in the first place.

I apologize if my comment sounded like pressure or even offensive, that wasn't my intention. I was backing up the point(s) raised by my colleagues with my own opinion, explaining it by bringing up some facts. 

> That's how it should work instead:

I'm not against that. That's actually a pretty sane pipeline. But that doesn't *possibly close* 2 things:

1. Again, that's not a *responsibility*. Nobody is ever *obligated* to contribute (including reports of undocumented behavior), or else we're in a cult. Welcome to the Minetest Church; All Heil the Mese.
2. The facts I stated earlier which you even backed up in your response. 3 VL releases is a lot, as we aren't rolling release or anything even close to it. I would suppose it's in ours (VL & MT) best interest to come to a resolution for this problem that would make everybody happy. If it would involve our behavior being left untouched, even if marked deprecated (which we will deal with the right next release), we'll be super happy. If it won't, we'll have to roll out a hot-patched version, which is not a bad thing as long as there'll be a valid way to achieve what we're trying to do (which is, invert colors).

To treat us differently or not is up to you, I just gave you the facts you already had.

**Comment by appgurueu** [comment]
> Again, that's not a responsibility. Nobody is ever obligated to contribute (including reports of undocumented behavior), or else we're in a cult. Welcome to the Minetest Church; All Heil the Mese. This doesn't mean we won't contribute, this just means nobody must do so (or else they'll be punished!).

I don't see where you're going with this "responsibility" argument. Because if we break it down that way all we end up at is "nobody is responsible for anything; we're all volunteers". By this argument we aren't responsible for maintaining backwards compatibility either.

Except we are. That's how "design by contract" (or whatever you want to call it) development works, and that's the only sane way it can work.

We assume responsibility for holding up our end of the contract - that is, we do our best to comply with what is outlined in the documentation. You assume responsibility for abiding by this documentation.

You have not held up your end of the contract here, and that results in more work for us - discussions like this, deprecating such features, implementing workarounds - and depending on how this is resolved, a weirder API for all future API users.

Not doing something - or doing something differently from what you expect - is also not "punishment".

We have compromised in the past and will compromise in the future, supporting some undocumented features just to limit the impact for users. But this is not something mod and game authors should come to expect.

So yes, I agree with Zughy here. When you get a PR which uses an undocumented feature and it "looks like it works", open an issue or PR, then wait for the feature to make it into Minetest, and only once that has happened merge the PR.

Everything else is just creating unnecessary work for everyone involved.

<details>
<summary>Coming back to your original message</summary>

I think you're greatly misrepresenting matters there, which is what prompted these responses:

> Breaking it means breaking backwards compatibility for a substantial amount of players.

As far as we are concerned, we are keeping backwards compatibility. VoxeLibre is relying on undefined behavior.

> There's nothing surprising about negative values acting weirdly in places they aren't supposed to be put in. Users might expect something different to happen with negative values. => Documenting this behavior should be a welcome change for everybody.

This is a massive non sequitur. Magic negative values are bad API design. Yes, we have them in a few places (for example "negative values represent percentages of screen size" in HUD elements). If you're reading `saturation = -0.5`, what do you think that does? Now, if you google negative saturation, what do you find? It makes no sense.

Key word is "weirdly". No, you wouldn't expect that negative saturation does "something something somewhat like inverting colors, but not quite". This PR doesn't even document it properly (see my comment).

So no, just documenting this as-is is definitely not "a welcome change for everybody". Zughy is 100% right that this would be weird API design.

> "There is no breaking compatibility as it's not documented" is like "There is no burglary as you never told me you had $10,000 in cash". Poor documentation (or lack of it) shouldn't be blamed on the user.

This is frankly ridiculous. Yes, Minetest **is not breaking compatibility** here. Calling this "poor documentation" is a massive misnomer. I think Desour's comment sums this up well.
</details>

**Comment by nauta-turbidus** [comment]
> So, nausea was added in VL 3 months ago, without reporting the undocumented behaviour https://git.minetest.land/VoxeLibre/VoxeLibre/commit/aa4ff1827d01a2466998ceba9da3249135e6b88a
> 
> after that, 3 VL releases have seen the light https://git.minetest.land/VoxeLibre/VoxeLibre/tags
> 
> And now, when we tell you that it might not be worth it, some of you try to pin it on Minetest, using the "VL is popular" card . Let me be very clear: that's not how it works and I personally don't tolerate this attempt of putting pressure on us when you didn't look for a dialogue in the first place. That's how it should work instead:
> 
>     1. people notice something undocumented
> 
>     2. to be sure, people open an issue to upstream (MT) to at least discuss this thing
> 
>     3. thing is discussed in upstream (VL  & MT)
> 
>     4. possibly a solution is found and everyone is happy: also future modders, who'll benefit from the cooperation between VL and MT
> 
> 
> That's what I do with my mods and my server, so I don't see why you can't do it as well. It allows the whole community to benefit from the work of others

The actual situation was:
1. I accidentally found out that's happening, didn't think about whether it is documented or not.
2. After the recent discussion on IRC, which you probably remember, I noticed it's undocumented and made this PR.
3. After your initial response, I linked this PR on our Discord to get comments from other devs. I'm not responsible for the content of their comments.

**Comment by nauta-turbidus** [comment]
> I'm more of a fan of the latter, as e.g. saturation and colour inversion wouldn't work (as we're using the same parameter)

Wdym? Saturation does work, i.e. -10 is very saturated aside of being inverted, and -0.1 is very not saturated but still inverted. Well, "inverted".

**Comment by Zughy** [comment]
@nauta-turbidus sorry I meant together

**Comment by ancientmarinerdev** [comment]
> It may be unintentional, but it's already in use. So, I see four main options:
> 
>     1. Document the existing behavior (this PR).
> 
>     2. Create a new parameter that inverts colors, and document that instead.
> 
>     3. Do nothing. The existing behavior remains undocumented.
> 
>     4. Remove the behavior entirely, breaking existing game behavior and provide no replacement.

I believe Technomunk has accurately give the options. 1 - 3 would be good options to VL based on the fact we're using this.

IMHO, 4 is outright hostile. Rather than doing nothing that takes zero effort (3), suggesting to remove it (4) is going out of the way to break a feature used actively in a game. I would understand if this behaviour was preventing correct behaviour causing side effects on other games or blocking a correct implementation of an upcoming feature, but from what I can see, it's neither. I'll have probably grown old and crossed over before ambient light gets over the line for example.

The other concerning thing here is the assertion that VL is in the wrong for not reporting this. I'm sorry, but before I took on maintainership in Dec 2022, I was told that Minetest were generally hostile to MCL2 at the time. I didn't believe it. I thought maybe we'd rubbed the project up the wrong way and there was a way to rebuild relationships. I later realised I was massively naive, and I have seen this one numerous occassions.

If you want people to raise bugs, you have to creation a culture, a safe space for people to raise bugs. If you punish people for raising bugs, you cannot in any way blame folk for not raising bugs. I have seen this same concern from other people working on other projects. Some weren't exactly comfortable volunteering this point. I will thank SmallJoker for being open minded here. It is appreciated to see that. However, you must realise, the reception an issue triager gets is VERY different from anyone else.

One criticism I have seen in the past of Minetest is they have actively kept bugs in because it would break mods and backwards compatibility is king here. To go out of your way to actively break a VL feature which will benefit no-one else, and have no other option in place would demonstrate that it's not about backwards compatibility and not breaking mods, it's about not breaking the sacred MTG. A game which is weirdly in maintenance mode, but is first class citizen in architectural decisions.

I've personally stepped down from VL, and do not really care one way or another which path is taken. I was working on a side project in MT, and was thinking of getting involved in another project and for the second time toxic positivity from Zughy is making me question whether it's worth it. I get that you're passionate about Minetest and that's cool, but when you're actively going after anyone that has even a slight concern about it, it's a bit too far. It doesn't help that others are quite keen to support what looks like a vendetta Zughy seems to have against VL. Maybe Zughy is the one bold enough to put down what others are feeling and do not really want to say. To go through commit history to prove it's new just to justify breaking it is disturbing. I guess this issue as an opportunity to keep satisfying that animosity.

"In short, my proposals as a modder, before it turns into a dumpfire"

Of course, to light the dumpster on fire, and then throw that out so people do not get a chance to respond to your hostile actions is not a good look. You're a representative of the project and have an obligation to be impartial. An engine that is build for all games and mods I would hope. I don't think it is responsible to appoint yourself judge and jury of what should be supported and what shouldn't. I'd like to hope you can triage with an open mind, that is if you really do want people to raise issues.

**Comment by Zughy** [comment]
@ancientmarinerdev fine

> and for the second time toxic positivity from Zughy is making me question whether it's worth it

Dude... I was the only staff member who listened to your concerns a few months ago, actually doing something about it https://forum.minetest.net/viewtopic.php?p=430579#p430579 . I don't know, I feel like this treatment is pretty mean. Whole paragraphs to portrait me as some sort of monster, I'm... honestly tired. You speak about a hypothetical vendetta, when I showcase VL in every convention I bring Minetest in, then I get blamed for digging through commit history to prove my point, and finally I get accused of not giving people the chance to respond to my "hostile actions" when they had literally answered before I published the comment.  

Fuck this shit, man. And fuck you, at this point. 4 AM trying to justify what I do because someone has decided to start the actual dumpfire, to then lecture me about how we're not welcoming enough and how I should behave instead, when you perfectly know I've been doing my best since day one to try to make Minetest development more open to all. The fact I got annoyed after two impartial comments doesn't justify all this. Unsubscribing for my mental health

**Comment by lhofhansl** [comment]
Let's not turn this into a big issue. Core devs make the decisions about the Core engine, and of course we're doing this with the modding community in mind.

Facts are:
1. This behavior is undocumented and I - as the person who put this in, finishing an abandoned patch, and who made it server controllable - did *not* intent this. Negative values are an oversight.
2. The unintended part seems to work fine.
3. The unintended part is already used.

I really do not care. If we decide to disable this, I'll happy do a PR for that.
If we decide to document this, then let's do that.

That all said... My vote would be to keep it as is and document it.

**Comment by rubenwardy** [comment]
@ancientmarinerdev Please can we avoid inflammatory comments and personal attacks. Portraying disagreement as if it is an attack is toxic and leads to people being less likely to raise issues. To restate: just because someone disagrees doesn't mean they're attacking you

Zughy was one of the main encouragers behind debundling MTG and has spoken against it many times, to say he has some bias for it is absurd.

Online communication is hard, please can we first assume good faith and realise that everyone is working to the same goal, to make Minetest better

**Comment by rubenwardy** [comment]
My personal opinion: it depends on the implementation of the formula. We should not be introducing weird APIs that were never designed. So if the negative doesn't act as one might expect, either the implementation should be changed and documented, or bounds checking and a new API added.

I would expect negative numbers to invert the colour and then apply a saturation of `abs(saturation)`

**Comment by ancientmarinerdev** [comment]
To mark my post as abusive I think is a little unfair, and to see my points misrepresented and some strawman arguments maybe feels like hiding issues is a good way not engage with them. I have gone through my post, and removed any references to anyone, and so it is purely focussed on my points. It's important and relevent to this issue:

> It may be unintentional, but it's already in use. So, I see four main options:
> 
>     1. Document the existing behavior (this PR).
> 
>     2. Create a new parameter that inverts colors, and document that instead.
> 
>     3. Do nothing. The existing behavior remains undocumented.
> 
>     4. Remove the behavior entirely, breaking existing game behavior and provide no replacement.

I believe Technomunk has accurately give the options. 1 - 3 would be good options to VL based on the fact we're using this.

IMHO, 4 is outright hostile. Rather than doing nothing that takes zero effort (3), suggesting to remove it (4) is going out of the way to break a feature used actively in a game. I would understand if this behaviour was preventing correct behaviour causing side effects on other games or blocking a correct implementation of an upcoming feature, but from what I can see, it's neither. I'll have probably grown old and crossed over before ambient light gets over the line for example.

The other concerning thing here is the assertion that VL is in the wrong for not reporting this. I'm sorry, but before I took on maintainership in Dec 2022, I was told that Minetest were generally hostile to MCL2 at the time. I didn't believe it. I thought maybe we'd rubbed the project up the wrong way and there was a way to rebuild relationships. I later realised I was massively naive, and I have seen this on numerous occassions.

If you want people to raise bugs, you have to create a culture, a safe space for people to raise bugs. If you punish people for raising bugs, you cannot in any way blame folk for not raising bugs. I have seen this same concern from other people working on other projects. Some weren't exactly comfortable volunteering this point. One example is below:

https://forum.minetest.net/viewtopic.php?p=430623#p430623 (thanks for the link, and the whole discussion is as valid as ever)

The project can be a little cliquey and if you get on the wrong side of that, it isn't a good experience. I will thank SmallJoker for being open minded here. It is appreciated to see that. However, you must realise, the reception an issue triager would get is VERY different from anyone else.

One criticism I have seen in the past of Minetest is they have actively kept bugs in because it would break mods and backwards compatibility is king here. To go out of your way to actively break a VL feature which will benefit no-one else, and have no other option in place would demonstrate that it's not about backwards compatibility and not breaking mods, it's about not breaking the sacred MTG. A game which is weirdly in maintenance mode, but is first class citizen in architectural decisions.

I've personally stepped down from VL, and do not really care one way or another which path is taken. I was working on a side project in MT, and was thinking of getting involved in another project and the more I see the reception to PRs and issues on the Minetest issue tracker, the more I question if that's a good investment of time.

To go through commit history to prove it's new just to justify breaking it is disturbing. 

"In short, my proposals as a modder, before it turns into a dumpfire"

I do not think comments like this are helpful. Minetest is an engine that is build for all games and mods I would hope. Anyone working on the issue tracker should be asking "what is this project trying to achieve?". "Is there a way the engine does this, or could support this?". I don't see other game engines taking time to tell game devs they're wrong for wanting something. Modders and game devs should be free to be creative and express themselves to create the experiences they visualise. If Minetest want people to only use things in the way intended, players will have the chance to choose from MTG, or MTG (with slight differences), or MTG (with other slight differences). That would be pretty dull. I'd like to hope triaging is done with an open mind, if Minetest genuinely want people to raise issues.

**Comment by ancientmarinerdev** [comment]
> Zughy was one of the main encouragers behind debundling MTG and has spoken against it many times, to say he has some bias for it is absurd.

A misrepresented point. One that I did make, a strawman argument.

> @ancientmarinerdev Please avoid inflammatory comments and personal attacks. Portraying disagreement as if it is an attack is toxic and leads to a less open discussion with people being less likely to raise issues. To restate: just because someone disagrees doesn't mean they're attacking you
> 
> Online communication is hard, please can we first assume good faith and realise that everyone is working to the same goal, to make Minetest better

I always try to assume good faith. It's just unfortunately since I got involved in Minetest in October 2022, and in particular as a maintainer since Dec 2022 there have been attacks for months on my character. What's worse, is that drama wasn't kept out of the Minetest discord or irc. It was actively encouraged and discussed (via positive emoji's from staff and no stamping down on it) bi-weekly to weekly.

A member of minetest staff accused me of being involved in a hostile takeover. Simply for putting my name forward when a project needed a maintainer. A conversation with a different person (a core dev) went as follows:

ancientmariner_ "i would rather it just wasn't discussed, or put in a blameless way "conflicts and disagreements""

core dev "I think the mcl2 teams puts themselves in a bad light, the issue tracker should have never gotten that toxic"

sfan5 "
<Nicu> and I find it pretty annoying that the Minetest developers don't care to look into the details but pass on their own (wrong) interpretation as valid
fyi i'm pretty sure (core dev) is speaking his personal opinion here, not in his position as a Minetest developer

I can be frank now I've stepped down, I can be open that my mental health took an absolute hammering, and it's taken a long time to even get it to a reasonable point (nowhere near where it was at the start). It wasn't personal attacks from people that actively disliked me. I expected that. It was the actions of those in official position on the Minetest project not stamping it down at all and in fact encouraging it. People involved include 2 discord mods, a core dev and a person regularly involved in PRs who was highly respected, and another 2/3 community members. Only 1 of those people owned that. Hint, it wasn't a person who was a mod or a core dev.

To clarify further, the person who is the core dev in that conversation was you, Rubenwardy. Actively passing verdicts on something you had limited knowledge of and helping to ensure that viewpoint cemented in many staff actively involved. To be honest, I only ever felt an improvement in culture since Desour, Sriqi, Lars were promoted. Minetest was toxic for a long time, it became a little less toxic when they stamped down on some toxicity on the Discord, and a little more in the last 6 months. Credit to GreenXenith for taking feedback on board.

For you to try and lecture me on taking things in good faith, when you do not express the same courtesy is astounding. To expect someone who was in a pressure cooker for 6-8 months, and to see regular personal attacks without being able to respond without multiple folk (including staff jumping in to support and enable those character attacks) jumping in to keep it going without warning.

I will say, getting kicked for that long, pretend it doesn't bother you, mask (because not everyone is neurotypical), and then engage to get lectures on behaviour from those have had actively roles in toxicity and haven't owned it.

File: doc/lua_api.md
Comment (appgurueu) [inline doc/lua_api.md:8488]: nitpick: 0, 1 are exclusive. not sure which notation we want to use for this. `(0, 1)` might be misunderstood as a tuple, `]0, 1[` is ugly and somewhat uncommon, `> 0 and < 1` is a bit verbose.

**Review by appgurueu — COMMENTED** [review]
The formula itself is pretty robust, so I don't think it's a big problem to document. While I'm not a big fan of "abusing" negative saturation like this, I can live with it.

I have one remaining concern however: How robust and consistent is this across different graphics configurations? It seems to me that we might get negative RGB components with a negative saturation, and I have no idea how we (and OGL) handle (and should handle) that.

**Comment by nauta-turbidus** [comment]
> The formula itself is pretty robust, so I don't think it's a big problem to document. While I'm not a big fan of "abusing" negative saturation like this, I can live with it.
> 
> I have one remaining concern however: How robust and consistent is this across different graphics configurations? It seems to me that we might get negative RGB components with a negative saturation, and I have no idea how we (and OGL) handle (and should handle) that.

https://community.khronos.org/t/texture-negative-values/71384/4

This seems to say that if we use actual floating-point output, it should work? Probably shouldn't completely fall apart anyway.

We could just clamp the value at the end of the shader, if we are ultimately unsure. Very high positive saturation may cause out-of-bounds outcomes as well, both over 1 and negative.

File: doc/lua_api.md
Comment (nauta-turbidus) [reply doc/lua_api.md]: Changed that to "values between 0 and 1". Will be clear at least. Clearer than `> 0 and < 1` IMHO, even.

**Comment by lhofhansl** [comment]
@nauta-turbidus due to the heated discussion here - not you but some of the other folks, it might be best to close this one and open a new one.

It's a bit weird that it escalated so quickly, we're not enemies here. :)

**Comment by nauta-turbidus** [comment]
> @nauta-turbidus due to the heated discussion here - not you but some of the other folks, it might be best to close this one and open a new one.
> 
> It's a bit weird that it escalated so quickly, we're not enemies here. :)

It's calm again though, I don't see a reason to close and reopen really.

**Comment by Zughy** [comment]
I've looked at the new version but I still find this PR a convoluted way to justify something that it shouldn't belong to saturation at all. Now documentation is complicated and if I hadn't followed the PR I wouldn't have understood anything

File: doc/lua_api.md
Comment (SmallJoker) [inline doc/lua_api.md:8553]: 1. Line limited to < 95 characters (according to https://dev.minetest.net/Code_style_guidelines#Don.27t_let_things_get_too_large )
2. More compact description (debatable)

```suggestion
          * A value of 0 corresponds to greyscale, and > 1 to oversaturated.
          * values < 0 cause an effect similar to inversion, but keeping original luma and
            being symmetrical in terms of saturation
            (e.g. -1 and 1 result in the same saturation and luma, but with different hues)
```

Comment (nauta-turbidus) [reply doc/lua_api.md:8553]: You're not including the `0 < saturation < 1` case though (nor the default 1, although that may be considered implicit).

Comment (nauta-turbidus) [reply doc/lua_api.md:8553]: Also I'm pretty sure I've seen longer lines like that in markdown files here.

Comment (SmallJoker) [reply doc/lua_api.md:8553]: >  although that may be considered implicit

I'd consider that common sense. Though that does not mean that I'm correct with this assumption.

> Also I'm pretty sure I've seen longer lines like that in markdown files here.

The C++ side does sometimes not follow the guidelines either, but changes that touch such lines are expected to correct it to provide a more uniform appearance.

Comment (nauta-turbidus) [reply doc/lua_api.md]: I wrote something similar to your suggestion. See if it's good enough.

**Review by SmallJoker — APPROVED** [review]
LGTM

**Review by appgurueu — APPROVED** [review]
> https://community.khronos.org/t/texture-negative-values/71384/4
> This seems to say that if we use actual floating-point output, it should work? Probably shouldn't completely fall apart anyway.

This seems to be in a bit of a different context (using shaders to do some computations, not to display stuff to the end user). It isn't quite clear to me how negative RGB values would get turned into the normalized RGB your screen uses.

I played around with this a bit. Since this is a nice effect, I'm approving on the condition that some level of consistency is given, otherwise we'll have to look into clamping or similar. Here's MTG silver sand biome with a saturation of `-10`, next to a snowy biome with blueberry bushes:

Screenshot from 2024-09-12 00-57-45

Do you see similar colors?

**Comment by nauta-turbidus** [comment]
@appgurueu 
obraz

**Pull request merged**
